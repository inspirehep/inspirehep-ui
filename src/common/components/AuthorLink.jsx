import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { Tooltip } from 'antd';

import ExternalLink from './ExternalLink';
import AuthorAffiliationList from './AuthorAffiliationList';

class AuthorLink extends Component {
  getAuthorHref() {
    const { author, recordId } = this.props;
    let href = `//inspirehep.net/author/profile/${author.get('full_name')}`;
    if (recordId != null) {
      href = `${href}?recid=${recordId}`;
    }
    return href;
  }

  getFullName() {
    const { author } = this.props;
    if (author.has('first_name')) {
      const firstName = author.get('first_name');
      const lastName = author.get('last_name', '');
      return `${firstName} ${lastName}`;
    }
    return author.get('full_name');
  }

  renderEditorSuffix() {
    const { author } = this.props;
    const roles = author.get('inspire_roles', []);

    if (roles.includes('editor')) {
      return <Tooltip title="editor">(ed.)</Tooltip>;
    }
    return null;
  }

  renderAffiliationsList() {
    const { author } = this.props;
    const affiliations = author.get('affiliations');
    return (
      affiliations && (
        <span className="pl1">
          (
          <AuthorAffiliationList
            affiliations={affiliations.map(affiliation =>
              affiliation.get('value')
            )}
          />
          )
        </span>
      )
    );
  }

  render() {
    const authorHref = this.getAuthorHref();
    return (
      <div className="di">
        <ExternalLink href={authorHref}>{this.getFullName()}</ExternalLink>
        {this.renderAffiliationsList()}
        {this.renderEditorSuffix()}
      </div>
    );
  }
}

AuthorLink.propTypes = {
  author: PropTypes.instanceOf(Map).isRequired,
  recordId: PropTypes.number,
};

AuthorLink.defaultProps = {
  recordId: undefined,
};

export default AuthorLink;
