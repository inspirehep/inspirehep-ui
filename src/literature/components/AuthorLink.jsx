import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';

class AuthorLink extends Component {
  getAuthorHref() {
    const { author, recordId } = this.props;
    let href = `//inspirehep.net/author/profile/${author.get('full_name')}`;
    if (recordId != null) {
      href = `${href}?recid=${recordId}`;
    }
    return href;
  }

  getAffiliationHref() {
    const { author } = this.props;
    const affiliation = author.getIn(['affiliations', 0, 'value']);
    return `//inspirehep.net/search?cc=Institutions&p=institution:"${affiliation}"`;
  }

  getFullName() {
    const { author } = this.props;
    if (author.has('first_name')) {
      const firstName = author.get('first_name');
      const lastName = author.get('last_name', '');
      return `${firstName} ${lastName}`;
    }
    return author.get('full_name');
  }

  renderAffiliationLink() {
    const { author } = this.props;
    const affiliation = author.getIn(['affiliations', 0, 'value']);
    if (affiliation) {
      const affiliationHref = this.getAffiliationHref();
      return (
        <span className="pl1 grey">
          (
          <a target="_blank" href={affiliationHref}>
            {affiliation}
          </a>
          )
        </span>
      );
    }
    return null;
  }

  render() {
    const authorHref = this.getAuthorHref();

    return (
      <div className="di">
        <a target="_blank" href={authorHref}>
          {this.getFullName()}
        </a>
        {this.renderAffiliationLink()}
      </div>
    );
  }
}

AuthorLink.propTypes = {
  author: PropTypes.instanceOf(Map).isRequired,
  recordId: PropTypes.number,
};

AuthorLink.defaultProps = {
  recordId: undefined,
};

export default AuthorLink;
