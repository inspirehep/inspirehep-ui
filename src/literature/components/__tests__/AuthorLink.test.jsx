import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import AuthorLink from '../AuthorLink';

describe('AuthorLink', () => {
  it('renders with full author', () => {
    const author = fromJS({
      full_name: 'Name, Full',
      first_name: 'Full',
      last_name: 'Name',
      affiliations: [
        {
          value: 'Affiliation',
        },
      ],
    });
    const wrapper = shallow(<AuthorLink author={author} recordId={12345} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders full author without recordId', () => {
    const author = fromJS({
      full_name: 'Name, Full',
      first_name: 'Full',
      last_name: 'Name',
      affiliations: [
        {
          value: 'Affiliation',
        },
      ],
    });
    const wrapper = shallow(<AuthorLink author={author} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders full author without last_name', () => {
    const author = fromJS({
      full_name: 'Name Full',
      first_name: 'Name Full',
      affiliations: [
        {
          value: 'Affiliation',
        },
      ],
    });
    const wrapper = shallow(<AuthorLink author={author} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders author with only name fields', () => {
    const author = fromJS({
      full_name: 'Name, Full',
      first_name: 'Full',
      last_name: 'Name',
    });
    const wrapper = shallow(<AuthorLink recordId={12345} author={author} />);
    expect(wrapper).toMatchSnapshot();
  });
});
