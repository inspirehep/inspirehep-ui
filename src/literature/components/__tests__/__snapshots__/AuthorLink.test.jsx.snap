// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorLink renders atuhor with only full_name 1`] = `
<div
  className="di"
>
  <a
    href="//inspirehep.net/author/profile/Name, Full?recid=12345"
    target="_blank"
  >
    Name, Full
  </a>
</div>
`;

exports[`AuthorLink renders full author without recordId 1`] = `
<div
  className="di"
>
  <a
    href="//inspirehep.net/author/profile/Name, Full"
    target="_blank"
  >
    Name, Full
  </a>
  <span
    className="pl1 grey"
  >
    (
    <a
      href="//inspirehep.net/search?cc=Institutions&p=institution:\\"Affiliation\\""
      target="_blank"
    >
      Affiliation
    </a>
    )
  </span>
</div>
`;

exports[`AuthorLink renders with full author 1`] = `
<div
  className="di"
>
  <a
    href="//inspirehep.net/author/profile/Name, Full?recid=12345"
    target="_blank"
  >
    Name, Full
  </a>
  <span
    className="pl1 grey"
  >
    (
    <a
      href="//inspirehep.net/search?cc=Institutions&p=institution:\\"Affiliation\\""
      target="_blank"
    >
      Affiliation
    </a>
    )
  </span>
</div>
`;
